### ISSUE TITLE
Oxigraph server `/store` uses a massive amount of RAM to import a small dataset

### ISSUE THREAD

**Issue by kelvinhammond** [oxigraph/oxigraph#844]
**Describe the bug**
Oxigraph server uses a massive amount of RAM to import a small dataset.
I have a dataset that is about 1 gigabyte in size in uncompressed text form. Oxigraph server's `/store` endpoint causes the server to use ~60G of RAM which it doesn't free after import. It also takes a decent amount of time to import this dataset.

```
> wc -l ./dataset.ttl 
15753057 ./dataset.ttl
```

**To Reproduce**
1. Which version of Oxigraph are you using? v0.3.22 On which platform? linux / docker
2. A command line or a code snippet that triggers the bug.

```
curl -T ./dataset.ttl -X POST -H 'Content-Type: application/turtle' 'http://localhost:7878/store'
```

**Comment by Tpt** [comment]
Thank you for trying Oxigraph and writing this bug report. By default the `/store` endpoint uses transactions, hence the memory consumption. You can either add the `no_transaction=1` URL query parameter or use the `oxigraph_server load` command to avoid transactions and get a much faster load speed and a way lower memory consumption.

But I anycase, not freeing memory after finishing loading the data is not good. I am going to investigate it.

**Comment by pvzhelnov** [comment]
Just reproduced a similar behavior using the `/update` endpoint. Set-up: `oxigraph/oxigraph:0.4`. Just under a million triples involved. The request occupied about 8 GiB of RAM and did not free it after completing. I reran the same request several times after that, and it stayed at about 8-9 GiB of RAM use as per `docker stats`. Not sure if this helps! I wonder what are ways to force Oxigraph to free memory after finishing to serve the transaction?

**Comment by Tpt** [comment]
> Just reproduced a similar behavior using the `/update` endpoint. Set-up: `oxigraph/oxigraph:0.4`. Just under a million triples involved. The request occupied about 8 GiB of RAM and did not free it after completing. I reran the same request several times after that, and it stayed at about 8-9 GiB of RAM use as per `docker stats`. Not sure if this helps! I wonder what are ways to force Oxigraph to free memory after finishing to serve the transaction?

Just to make sure: Are you using the disk storage (with the `--location` parameter) or the default in memory storage? If you use the disk storage it's definitely not an expected behavior and looks like a memory leak or something similar.

**Comment by pvzhelnov** [comment]
> > Just reproduced a similar behavior using the `/update` endpoint. Set-up: `oxigraph/oxigraph:0.4`. Just under a million triples involved. The request occupied about 8 GiB of RAM and did not free it after completing. I reran the same request several times after that, and it stayed at about 8-9 GiB of RAM use as per `docker stats`. Not sure if this helps! I wonder what are ways to force Oxigraph to free memory after finishing to serve the transaction?
> 
> Just to make sure: Are you using the disk storage (with the `--location` parameter) or the default in memory storage? If you use the disk storage it's definitely not an expected behavior and looks like a memory leak or something similar.

Yes, I am serving Oxigraph using `serve --location /dir`. At first glance, do you think it is more likely to be due to the loops of virtualization involved (e.g., Docker under lima) rather than an issue with Oxigraph?

**Comment by Tpt** [comment]
> > > Just reproduced a similar behavior using the `/update` endpoint. Set-up: `oxigraph/oxigraph:0.4`. Just under a million triples involved. The request occupied about 8 GiB of RAM and did not free it after completing. I reran the same request several times after that, and it stayed at about 8-9 GiB of RAM use as per `docker stats`. Not sure if this helps! I wonder what are ways to force Oxigraph to free memory after finishing to serve the transaction?
> > 
> > 
> > Just to make sure: Are you using the disk storage (with the `--location` parameter) or the default in memory storage? If you use the disk storage it's definitely not an expected behavior and looks like a memory leak or something similar.
> 
> Yes, I am serving Oxigraph using `serve --location /dir`. At first glance, do you think it is more likely to be due to the layers of virtualization involved (e.g., Docker under lima) rather than an issue with Oxigraph?
> 
> By the way, restarting the container does free the RAM in my case.

I really don't know. I will try to reproduce this locally.

**Comment by pvzhelnov** [comment]
> I really don't know. I will try to reproduce this locally.

Many thanks. If it helps, here is the SPARQL query:

```sparql
DELETE {
  GRAPH ?old_g {
    ?old_s ?old_p ?old_o
  }
}
INSERT {
  GRAPH ?new_g {
    ?new_s ?new_p ?new_o
  }
}
WHERE {
  GRAPH ?old_g {
    ?old_s ?old_p ?old_o
  }
  
  BIND(
    IF(STRSTARTS(STR(?old_g), "https://example.com"),
       IRI(REPLACE(STR(?old_g), "https://example.com", "https://example1.com")),
       ?old_g
    ) AS ?new_g)
  
  BIND(
    IF(STRSTARTS(STR(?old_s), "https://example.com"),
       IRI(REPLACE(STR(?old_s), "https://example.com", "https://example1.com")),
       ?old_s
    ) AS ?new_s)
  
  BIND(
    IF(STRSTARTS(STR(?old_p), "https://example.com"),
       IRI(REPLACE(STR(?old_p), "https://example.com", "https://example1.com")),
       ?old_p
    ) AS ?new_p)
  
  BIND(
    IF(ISIRI(?old_o) && STRSTARTS(STR(?old_o), "https://example.com"),
       IRI(REPLACE(STR(?old_o), "https://example.com", "https://example1.com")),
       ?old_o
    ) AS ?new_o)
}
```

```json
{
  "contentType": null,
  "status": 204,
  "executionTime": 86618
}
```

And some idea about my dataset:

```sparql
SELECT (COUNT(*) as ?g) WHERE {
  GRAPH ?g {?s ?p ?o . }
}
```

```json
{
  "head": {
    "vars": [
      "g"
    ]
  },
  "results": {
    "bindings": [
      {
        "g": {
          "type": "literal",
          "value": "963040",
          "datatype": "http://www.w3.org/2001/XMLSchema#integer"
        }
      }
    ]
  }
}
```

**Comment by pvzhelnov** [comment]
> At first glance, do you think it is more likely to be due to the layers of virtualization involved (e.g., Docker under lima) rather than an issue with Oxigraph?

Confirming this is unlikely to be related to the lima/containerd setup on macOS as I just reproduced this behavior exactly on native Debian 12 (bookworm) – still under Docker, however (same image, setup, and dataset as above).

Also, the issue with RAM remaining occupied reproduces for me for all SPARQL queries, including SELECT on `/query` endpoint. It basically looks like whichever query gets executed occupies some RAM and keeps that – until, with consecutive queries, RAM is overloaded and the container restarts.
